kernels/volk/volk_4ic_deinterleave_8i_x2.h

```diff
@@ -0,0 +1,127 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * GNU Radio is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3, or (at your option)
+ * any later version.
+ *
+ * GNU Radio is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNU Radio; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+/*!
+ * \page volk_4ic_deinterleave_8i_x2
+ *
+ * \b Overview
+ *
+ * Deinterleaves the complex 4-bit integer vector into I & Q vector data
+ * and converts them to 8-bit integers.
+ *
+ * <b> Dispatcher Prototype</b>
+ * \code
+ * void volk_4ic_deinterleave_8i_x2(int8_t* iBuffer, int8t_t* qBuffer, const int8_t*
+ * complexVector, unsigned int num_points) \endcode
+ *
+ * \b Inputs
+ * \li complexVector: The complex input vector.
+ * \li num_points: The number of complex data values to be deinterleaved.
+ *
+ * \b Outputs
+ * \li iBuffer: The I buffer output data.
+ * \li qBuffer: The Q buffer output data.
+ *
+ * \b Example
+ * \code
+ * int N = 10000;
+ *
+ * volk_4ic_deinterleave_8i_x2();
+ *
+ * volk_free(x);
+ * \endcode
+ *
+ */
+
+#ifndef INCLUDED_volk_4ic_deinterleave_8i_x2_a_H
+#define INCLUDED_volk_4ic_deinterleave_8i_x2_a_H
+
+#include <inttypes.h>
+#include <stdio.h>
+
+#ifdef LV_HAVE_GENERIC
+
+static inline void volk_4ic_deinterleave_8i_x2_generic(int8_t* iBuffer,
+                                                       int8_t* qBuffer,
+                                                       const int8_t* complexVector,
+                                                       unsigned int num_points)
+{
+    const int8_t* complexVectorPtr = complexVector;
+    int8_t* iBufferPtr = iBuffer;
+    int8_t* qBufferPtr = qBuffer;
+    for (unsigned int i = 0; i < num_points; i++) {
+        *iBufferPtr++ = (int8_t)(*complexVectorPtr) >> 4;
+        *qBufferPtr++ = ((int8_t)(*complexVectorPtr++) << 4) >> 4;
+    }
+}
```

Comment:
Unfortunately bitshifting negative numbers is undefined behaviour (or implementation defined, I can't remember) though most compilers will do what you expect.

The other problem here is that C/C++ only define integer arithmetic over `int` types, so smaller types are up-cast. So technically the result of

```c++
(int8_t)(*complexVectorPtr++) <<4
```

is of type `int`.

I would rewrite this as either:

```c
for (unsigned int i = 0; i < num_points; i++) {
        *iBufferPtr++ = (*complexVectorPtr >> 4);
        *qBufferPtr++ = (int8_t)((*complexVectorPtr++)  << 4) >> 4;
}
```
if you are happy with the undefined behaviour, or:

```c
for (unsigned int i = 0; i < num_points; i++) {
        *iBufferPtr++ = (*complexVectorPtr) /16;
        *qBufferPtr++ = (int8_t)((*complexVectorPtr++) *16) /16;
   }
```
if you are not. The latter might not optimise as well as the former, but is guaranteed to be correct. Note that I've left the last cast to `int8_t` as implicit in each case - I'm not sure if this agrees with your coding style.

Reply:
Just saw the use case @dkozel was working on, and both nibbles would have a sign bit.  I wonder if this may be the correct way.  That would put the signed bits where they need to be then scale the values down appropriately?

for (unsigned int i = 0; i < num_points; i++) {
        *iBufferPtr++ = (*complexVectorPtr & 0xF0)/16;
        *qBufferPtr++ = (int8_t)((*complexVectorPtr++)  << 4)/16;
}